### DISCUSSION TITLE
ES Module Support for config

### DISCUSSION THREAD

**Discussion by acalvino4** [tailwindlabs/tailwindcss#5858] [Ideas]
When my tailwind.config.js file is an ES Module, I get this below error while purging my css. This is a request to enable module support from tailwind.

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /var/www/craft/config/build/tailwind.config.js
require() of ES modules is not supported.
require() of /var/www/craft/config/build/tailwind.config.js from /var/www/craft/node_modules/tailwindcss/lib/cli.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename tailwind.config.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /var/www/craft/package.json.
```

**Comment by acalvino4** [comment]
I found a workaround - I'm using gulp, so I just loaded my config into a javascript variable, then passed it into the tailwind constructor, rather than passing the config path to tailwind.

I still think es6 module support would be awesome though
```
...
import twconfig from './tailwind.config.js'
...
function css() {
  return src('src/css/*.css')
    .pipe(concat('main.min.css'))
    .pipe(postcss([
      tailwind(twconfig),
      autoprefix(),
    ]))
    .pipe(clean())
    .pipe(dest('web/dist/css/'))
}
```

**Comment by Robert9655** [comment]
To explicitly start the agent, import elastic-apm-node and call the apm. Start method with the agent configuration object. Type scripted it. If you're using.
[hardee's survey
](https://www.tellhappystar.net/)

**Comment by michaelhays** [comment]
With Next.js [adding support for ESM projects](https://github.com/vercel/next.js/pull/33637), it's even more necessary now for Tailwind to support using an ESM config file.

I'm unable to generate my Tailwind classes using PostCSS in an ESM Next.js project, and have to generate it manually using the [Tailwind CLI](https://tailwindcss.com/docs/installation).

**Comment by vibegui** [comment]
This is also unfortunate for deno-based projects. Currently, I'm unable to import my `tailwind.config.js` into my deno project directly, having to go through this hop: https://github.com/FutureDrivenDev/fresh-init-tailwind/blob/main/utils/twind.ts#L6

Case in point: I'm actually using `twind` there under the hood, but the plugins like the excellent Tailwind CSS IntelliSense rely on CJS too, because of tailwind: https://github.com/tailwindlabs/tailwindcss-intellisense/issues/399#issuecomment-1059524925

**Comment by damianobarbati** [comment]
@michaelhays how did you solve this?

**Reply by michaelhays** [reply]
Sorry for the late response -- I haven't solved it. I reverted that Next.js project back to CJS, and imagine it'll have to stay there for at least another year, while waiting for the ecosystem to catch up and stabilize :upside_down_face:

**Comment by minht11** [comment]
Are there any plans on solving this? Ecosystem as a whole is migrating to ESM modules and the longer it isn't supported the more painful it is gonna be. [postcss-load-config](https://github.com/postcss/postcss-load-config/blob/main/CHANGELOG.md) supports loading ESM modules for over a year now, so any news from Tailwind team about status of this feature would be welcome.

**Comment by muradbu** [comment]
+1 for this. We have a monorepo that's mainly ESModules based.

**Comment by dionsaur84** [comment]
you can use the following workaround for now.

`postcss.config.js`:

```js
import autoprefixer from 'autoprefixer'
import tailwind from 'tailwindcss'
import tailwindConfig from './tailwind.config.js'

export default {
    plugins: [autoprefixer, tailwind(tailwindConfig)],
}
```

`tailwind.config.js`:

```js
export default {
    content: ['app/components/**/*.jsx'],
    theme: {
        extend: {},
    },
    plugins: [],
}
```

**Comment by ekwoka** [comment]
The officially supported way is to just make the config be `.cjs` like `tailwind.config.cjs`

You still won't be able to use ESM inside the file itself, but this will allow your project to be ESM, and the config be CJS.

**Reply by sindelio** [reply]
Works for me, however pretty ugly to do and see. +1 for es6 module support in the config file.

**Reply by michaelhays** [reply]
It'll be released soon: https://github.com/tailwindlabs/tailwindcss/pull/10785

**Reply by sindelio** [reply]
Glad to hear that!

**Reply by shaobeichen** [reply]
works for me

**Comment by acalvino4** [comment]
There have been some nice workarounds suggested, but at the core I think most people on this thread just feel that modern js should have first-class support in modern libraries (to be fair, tailwind is unfortunately far from the only library that fails at allowing esm config). Sure, its not too hard to make it work. It's just annoying when your whole project is in ts and you have to then be adding linter exceptions to allow cjs syntax, not getting config validation, or having to find threads like this to work out how to translate the docs config examples to esm.

I had been using the workaround I posted at the top here, but wanted to use the multiple config file approach introduced in 3.2 - no luck, I get "Cannot use import statement outside a module" (it assumes my config is not esm); so looks like i gotta move back to cjs...

**Comment by acalvino4** [comment]
@adamwathan I don't think this thread has had any response from a maintainer, and I think the number of upvotes (#3 in all-time top idea discussions) warrants at least some word - are there any plans for this feature in the roadmap, and if so, are we talking short-term or long-term?

**Reply by michaelhays** [reply]
Here's the latest response that I know of from the team: https://github.com/tailwindlabs/tailwindcss/pull/3544#issuecomment-1203217456

**Comment by acalvino4** [comment]
Anyone find a workaround for using esm config with new `@config` syntax?

**Comment by julioflima** [comment]
I found a workaround that resolved my problem. Maybe that could resolve your problem too.

I couldn't import `.js` files, not even changing to `.cjs`, actually changing the tailwind stopped working. I'm not using a major recent application (legacy).

So, I let all variables that I need to use in `theme.ts` inside of a `colors.json`. In the `tailwind.config.js` I could import files as `.json`, and that it was possible to import too inside of `theme.ts`.

![code](https://user-images.githubusercontent.com/17098382/224752249-5b652a50-f63f-43b9-bd93-a1d2fad7ae57.png)

**Comment by alvarlagerlof** [comment]
This is now fixed with https://tailwindcss.com/blog/tailwindcss-v3-3#esm-and-typescript-support
